File: apps/backend/src/agent/graph.py
# ============================================================================
# graph.py - SIMPLIFIED
# ============================================================================
"""Main graph construction without todo integration."""

from langgraph.graph import StateGraph, START, END
from typing import Literal

from agent.state import AgentState
from agent.config import AgentConfig
from agent.llm_loader import create_llm, LLMConfig
from agent.tools import get_all_tools
from agent.middleware.base import console
from agent.middleware.memory import MemoryLoaderMiddleware
from agent.middleware.summarization import SummarizationMiddleware
from agent.middleware.model_call import ModelCallMiddleware
from agent.middleware.tracking import (
    StartMiddleware,
    TokenTrackerMiddleware,
    PIIDetectionMiddleware,
    EndMiddleware,
)
from agent.middleware.tools import ToolNodeWithMiddleware

# Note: LangGraph API handles persistence automatically.
# For custom postgres connection, set POSTGRES_URI environment variable.
# See: https://langchain-ai.github.io/langgraph/cloud/reference/env_var/#postgres_uri_custom


# ============================================================================
# Conditional Edge Functions
# ============================================================================


def should_continue(state: AgentState) -> Literal["tools", "check_completion"]:
    """Determine routing after model call."""
    last_message = state["messages"][-1]
    model_calls = state.get("model_calls", 0)

    try:
        config = AgentConfig.from_env()
        max_calls = config.max_model_calls
    except:
        max_calls = 20

    if model_calls >= max_calls:
        console.log(f"[should_continue] Max model calls ({max_calls}) reached")
        return "check_completion"

    if hasattr(last_message, "tool_calls") and last_message.tool_calls:
        console.log(
            f"[should_continue] Model called {len(last_message.tool_calls)} tool(s)"
        )
        return "tools"

    console.log(f"[should_continue] No tools called, checking completion")
    return "check_completion"


def check_completion_routing(state: AgentState) -> Literal["model", "EndMiddleware"]:
    """Check if agent should continue or end."""
    model_calls = state.get("model_calls", 0)

    try:
        config = AgentConfig.from_env()
        max_calls = config.max_model_calls
    except:
        max_calls = 20

    # 1. HARD STOP: Check limits first
    if model_calls >= max_calls:
        console.log(f"[check_completion] Max calls reached, ending")
        return "EndMiddleware"

    # 2. STANDARD CHAT LOGIC
    # If no tools were called (why we are here), the LLM has finished its thought.
    console.log(f"[check_completion] Standard chat response complete. Ending.")
    return "EndMiddleware"


# ============================================================================
# Graph Builder
# ============================================================================


def create_llm_from_config(config: AgentConfig):
    """Create an LLM instance from AgentConfig."""
    # Get model settings from model_config if available
    if config.model_config is not None:
        model_cfg = config.model_config
        return create_llm(
            provider=model_cfg.provider.value if model_cfg.provider else None,
            model_name=model_cfg.model_name,
            temperature=model_cfg.temperature,
            max_tokens=model_cfg.max_tokens,
            base_url=model_cfg.base_url,
        )

    # Fallback for backward compatibility - provider is already a string from enum
    return create_llm(
        provider=config.provider.value if config.provider else None,
        model_name=config.model_name,
        temperature=config.temperature,
        max_tokens=config.max_tokens,
        base_url=config.base_url,
    )


def build_graph(config: AgentConfig = None):
    """Build the agent graph."""
    config = config or AgentConfig.from_env()
    tools = get_all_tools()

    llm = create_llm_from_config(config)
    workflow = StateGraph(AgentState)

    # ========================================================================
    # NODE DEFINITIONS
    # ========================================================================
    workflow.add_node("StartMiddleware", StartMiddleware())

    if config.enable_memory_loader:
        workflow.add_node("memory", MemoryLoaderMiddleware())

    if config.enable_summarization:
        workflow.add_node(
            "summarize", SummarizationMiddleware(config.summarization_threshold, config)
        )

    workflow.add_node("model", ModelCallMiddleware(llm, tools, config))

    if config.enable_token_tracking:
        workflow.add_node("token_tracker", TokenTrackerMiddleware())

    workflow.add_node("tools", ToolNodeWithMiddleware(tools))
    workflow.add_node("check_completion", lambda state: {})

    if config.enable_pii_detection:
        workflow.add_node("pii_detect", PIIDetectionMiddleware())

    workflow.add_node("EndMiddleware", EndMiddleware())

    # ========================================================================
    # LINEAR EDGES - Initial setup flow
    # ========================================================================
    workflow.add_edge(START, "StartMiddleware")

    current = "StartMiddleware"
    if config.enable_memory_loader:
        workflow.add_edge(current, "memory")
        current = "memory"

    if config.enable_summarization:
        workflow.add_edge(current, "summarize")
        current = "summarize"

    workflow.add_edge(current, "model")

    # ========================================================================
    # MAIN LOOP - Model -> Token Tracker -> Routing
    # ========================================================================

    # After model, decide what to do
    if config.enable_token_tracking:
        # Model -> Token Tracker (always)
        workflow.add_edge("model", "token_tracker")

        # Token Tracker -> Conditional routing
        workflow.add_conditional_edges(
            "token_tracker",
            should_continue,
            {"tools": "tools", "check_completion": "check_completion"},
        )
    else:
        # No token tracking - direct conditional from model
        workflow.add_conditional_edges(
            "model",
            should_continue,
            {"tools": "tools", "check_completion": "check_completion"},
        )

    # Tools always goes back to model to process results
    workflow.add_edge("tools", "model")

    # ========================================================================
    # COMPLETION ROUTING
    # ========================================================================
    workflow.add_conditional_edges(
        "check_completion",
        check_completion_routing,
        {"model": "model", "EndMiddleware": "EndMiddleware"},
    )

    # ========================================================================
    # FINAL EXIT
    # ========================================================================
    if config.enable_pii_detection:
        workflow.add_edge("EndMiddleware", "pii_detect")
        workflow.add_edge("pii_detect", END)
    else:
        workflow.add_edge("EndMiddleware", END)

    return workflow.compile()


# Default graph instance - lazy loaded to avoid building at import time
_graph_instance = None

# Default tools - loaded once
_default_tools = None


def get_graph():
    """Get the default graph instance (lazy loaded)."""
    global _graph_instance
    if _graph_instance is None:
        _graph_instance = build_graph()
    return _graph_instance


def create_graph(tools=None):
    """Create a new graph instance with optional custom tools.

    Args:
        tools: Optional list of tools to use. If None, uses default tools.

    Returns:
        Compiled StateGraph instance.
    """
    from agent.config import AgentConfig
    from agent.tools import get_all_tools

    config = AgentConfig.from_env()
    all_tools = tools if tools is not None else get_all_tools()

    # Build graph with the specified tools
    workflow = StateGraph(AgentState)

    # Add nodes (simplified version)
    from agent.middleware.tracking import StartMiddleware, EndMiddleware
    from agent.middleware.memory import MemoryLoaderMiddleware
    from agent.middleware.summarization import SummarizationMiddleware
    from agent.middleware.model_call import ModelCallMiddleware
    from agent.middleware.tools import ToolNodeWithMiddleware

    workflow.add_node("StartMiddleware", StartMiddleware())

    if config.enable_memory_loader:
        workflow.add_node("memory", MemoryLoaderMiddleware())

    if config.enable_summarization:
        workflow.add_node(
            "summarize", SummarizationMiddleware(config.summarization_threshold, config)
        )

    workflow.add_node(
        "model", ModelCallMiddleware(create_llm_from_config(config), all_tools, config)
    )
    workflow.add_node("tools", ToolNodeWithMiddleware(all_tools))
    workflow.add_node("check_completion", lambda state: {})
    workflow.add_node("EndMiddleware", EndMiddleware())

    # Add edges
    workflow.add_edge(START, "StartMiddleware")
    workflow.add_edge("StartMiddleware", "model")
    workflow.add_edge("tools", "model")
    workflow.add_edge("EndMiddleware", END)

    return workflow.compile()


def __getattr__(name):
    """Lazy loading for graph to avoid building at import time."""
    global _graph_instance
    if name == "graph":
        if _graph_instance is None:
            _graph_instance = build_graph()
        return _graph_instance
    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")


def get_default_llm():
    """Get the default LLM using environment variables."""
    return create_llm()


create_llm_module = create_llm_from_config

# Default graph instance - eagerly loaded for LangGraph compatibility
graph = build_graph()

File: apps/backend/src/agent/graph.ts
import type { AIMessage } from "@langchain/core/messages";
import type { RunnableConfig } from "@langchain/core/runnables";
import { END, START, StateGraph } from "@langchain/langgraph";
import { FileSystemCheckpointer } from "./fs-checkpointer.js";
import { AgentNode } from "./nodes/agent.js";
import { ApprovalGate } from "./nodes/approval-gate.js";
import { EndMiddleware } from "./nodes/end-middleware.js";
import { initializeMemory, MemoryRetrieval } from "./nodes/memory-retrieval.js";
import { StartMiddleware } from "./nodes/start-middleware.js";
import { ToolExecution } from "./nodes/tool-execution.js";
import { type AgentState, AgentStateAnnotation } from "./state.js";

initializeMemory();

/**
 * Route after AgentNode: tool calls always go to ApprovalGate
 */
const routeAfterAgent = (state: AgentState): "ApprovalGate" | "EndMiddleware" => {
  const lastMessage = state.messages.at(-1);

  // No message or not AI message -> end
  if (!lastMessage || lastMessage._getType() !== "ai") {
    return "EndMiddleware";
  }

  const aiMessage = lastMessage as AIMessage;
  const toolCalls = aiMessage.tool_calls || [];

  // No tool calls -> end
  if (toolCalls.length === 0) {
    return "EndMiddleware";
  }

  // All tool calls go through ApprovalGate
  console.log("[Graph] Routing to ApprovalGate for tool processing");
  return "ApprovalGate";
};

/**
 * Route after ApprovalGate:
 * - If tools were rejected -> AgentNode (with ToolMessage feedback)
 * - If tools approved -> ToolExecution
 */
const routeAfterApproval = (state: AgentState): "ToolExecution" | "AgentNode" => {
  // If tools were rejected, route back to AgentNode with feedback
  if (state.tools_rejected) {
    console.log("[Graph] Tools rejected, routing to AgentNode with feedback");
    return "AgentNode";
  }

  // Tools approved, execute them
  return "ToolExecution";
};

/**
 * Check if we should continue the agent loop after ToolExecution
 */
const shouldContinue = (state: AgentState): "AgentNode" | "EndMiddleware" => {
  const envLimit = process.env.MAX_MODEL_CALLS ? parseInt(process.env.MAX_MODEL_CALLS, 10) : 50;
  const maxCalls = state.metadata?.max_model_calls || envLimit;

  if (state.model_calls >= maxCalls) {
    console.warn(`[Graph] Max model calls (${maxCalls}) reached`);
    return "EndMiddleware";
  }

  return "AgentNode";
};

export const graph = new StateGraph(AgentStateAnnotation)
  .addNode("StartMiddleware", StartMiddleware)
  .addNode("MemoryRetrieval", MemoryRetrieval)
  .addNode("AgentNode", AgentNode)
  .addNode("ApprovalGate", ApprovalGate)
  .addNode("ToolExecution", ToolExecution)
  .addNode("EndMiddleware", EndMiddleware)

  .addEdge(START, "StartMiddleware")
  .addEdge("StartMiddleware", "MemoryRetrieval")
  .addEdge("MemoryRetrieval", "AgentNode")

  // AgentNode -> ApprovalGate (if tools) or EndMiddleware (if no tools)
  .addConditionalEdges("AgentNode", routeAfterAgent, {
    ApprovalGate: "ApprovalGate",
    EndMiddleware: "EndMiddleware",
  })

  // ApprovalGate -> ToolExecution (approved) or AgentNode (rejected with feedback)
  .addConditionalEdges("ApprovalGate", routeAfterApproval, {
    ToolExecution: "ToolExecution",
    AgentNode: "AgentNode",
  })

  // ToolExecution -> AgentNode (continue) or EndMiddleware (max calls)
  .addConditionalEdges("ToolExecution", shouldContinue, {
    AgentNode: "AgentNode",
    EndMiddleware: "EndMiddleware",
  })

  .addEdge("EndMiddleware", END)

  .compile({ checkpointer: new FileSystemCheckpointer() });

console.log("[Graph] Simplified graph compiled:");
console.log("  Flow: START → Start → Memory → Agent → ApprovalGate → [Tools | Agent] → End → END");
console.log("  - All tool calls go through ApprovalGate");
console.log("  - Rejected tools return to AgentNode with ToolMessage feedback");
